### PR TITLE
Simplify setup.py, make CmdStan the default backend

### PR DESCRIPTION
**Pull request by WardBrian**
Working wheels (cmdstanpy only): https://github.com/WardBrian/prophet/actions/runs/1855227353

This WIP PR would be the next step for #2041. 

Currently it:
1. Makes the repackaging of CmdStan reproducible (doesn't assume an existing installation), and optional (can be disabled with environment variable `PROPHET_REPACKAGE_CMDSTAN=false`)
2. Checks to see if cmdstan was repackaged before setting the cmdstan path
2. Simplifies `setup.py` accordingly
3. Makes `CMDSTANPY` the default backend
4. Removes the distinction between the unix and windows model, as cmdstanpy can compile the "unix" model on all platforms

I'd like to test the current setup and wheel builds, hence this PR and the changes to the github actions files. I am still waiting to hear back from my company about the CLA, so this might sit as a draft until after the holidays.

Still yet to be done:
5. Refactor to remove PyStan entirely

### CODE REVIEW

**Comment by tcuongd** [comment]
Thanks for starting this Brian! Happy to help where I can. I've just merged in some changes to `setup.py`: https://github.com/facebook/prophet/pull/2089 that allow building on Windows using `cmdstanpy==0.9.68`. I'm conscious that we can simplify + refactor setup.py quite a bit in this PR, but at least it gives us a working example to compare against.

**Comment by WardBrian** [comment]
@tcuongd I've merged to try to maintain the changes in that PR, but alongside these repackaging changes.

Have you tested that it is not necessary to have the toolchain installed on end-user's machines? I'm not sure how much is statically linked

**Comment by WardBrian** [comment]
@tcuongd are you able to allow workflow runs for this PR so I can continue testing?

**Comment by tcuongd** [comment]
> Have you tested that it is not necessary to have the toolchain installed on end-user's machines? I'm not sure how much is statically linked

Hmm are you specifically talking about the Rtools toolchain on Windows, or more generally for Linux and MacOS as well?

In the `wheel.yml` workflow, we:

* Build the package in an isolated environment
* Then install the built wheel into the global python environment, and run the tests in the `tests/` folder.

The second step would just use all the pre-packaged code, so I _think_ if there's any issues these tests would fail. Would definitely appreciate a second look at this to confirm though.

The MacOS and Linux packages seem to work fine testing on my machines, but I'm not sure whether the Windows package would work for someone who doesn't have Rtools installed. We'd want to call out that prerequisite pretty clearly, although I don't think it needs to be automated as a part of the installation.

**Comment by WardBrian** [comment]
Yeah, specifically I wonder if someone on windows needs to have a Windows-compatible gcc/glibc. I can try to check once I get these wheels building

**Comment by WardBrian** [comment]
@tcuongd - I’m not very familiar with the test suite, could you help figure out why this test is failing? https://github.com/facebook/prophet/runs/4768342262?check_suite_focus=true

Otherwise we passed on Ubuntu. It looks like at least macOS failed because of a GitHub rate limit downloading cmdstan

**Comment by tcuongd** [comment]
> could you help figure out why this test is failing?
Yeah sure I can take a look. Really weird that it's failing though if we haven't changed any other code.

> GitHub rate limit downloading cmdstan
Ah yeah, I remember now, this is why I opted to download cmdstan via `wget` - it's much less fickle and faster as well.

**Comment by WardBrian** [comment]
For me, the same test fails on `main`

**Comment by WardBrian** [comment]
@tcuongd I'm happy to continue working on this if you get a chance to debug that test

File: .github/workflows/wheel.yml
Comment (akosfurton) [inline .github/workflows/wheel.yml:149]: I believe these lines are causing the rate limiting. They should not be removed and cmdstan should be downloaded via wget

**Comment by WardBrian** [comment]
I have not looked into them besides confirming they occur on `main` if I use the cmdstanpy backend.

File: .github/workflows/wheel.yml
Comment (WardBrian) [reply .github/workflows/wheel.yml:149]: I suspect you're correct here. Downloading via wget makes `setup.py` much more complicated if we care about doing the right thing for users who didn't pre-download cmdstan, though. I'll look into it, though I think the test failures are not related to this

**Comment by WardBrian** [comment]
This is the kind of error we have seen when multiple cmdstan processes are trying to write to the same file at the same time, usually in threaded environments where the chain ID is not being explicitly set. Does that sound like a plausible explanation for this error @malmashhadani-88 ?

**Comment by akosfurton** [comment]
@WardBrian , exactly as hypothesized.

The test is running multiple models in parallel (cross-validating the time series), so all of the output files were being written to the same location.

By adding `output_dir = mkdtemp(),` to line 96 of models.py, all tests appear to pass

**Comment by WardBrian** [comment]
@malmashhadani-88 glad it was a simple fix. Could you be more specific about what I should change in the test?

**Comment by WardBrian** [comment]
@akosfurton I believe I have made the change as you describe

File: .github/workflows/wheel.yml
Comment (akosfurton) [reply .github/workflows/wheel.yml:149]: ^ These don't appear to be an issue anymore. All CI builds fail on the same test

File: README.md
Comment (tcuongd) [inline README.md:69]: `NO_REPACKAGE_CMDSTAN=True` ?

File: README.md
Comment (tcuongd) [inline README.md:79]: We should add something here about being able to install from `pip` as well, but that will firstly require installing the `RTools` toolchain (e.g. via `cmdstanpy`).

File: python/prophet/models.py
Comment (tcuongd) [inline python/prophet/models.py:74]: Just wondering if you've tested the pip Windows install on your machine (you can download the wheel)? I added this because I was getting an error about TBB when trying to load the model.

File: python/requirements.txt
Comment (tcuongd) [inline python/requirements.txt:3]: If we're removing this should we catch import errors in parts of the code which use pystan? With a check for the version as well. Otherwise happy to leave this in for now and do that in another PR.

File: python/setup.py
Comment (tcuongd) [inline python/setup.py:27]: Yeah I'm not sure what the history is behind having different model code for Windows, but if the compilation and tests pass on Windows then happy to make this simplifying change.

File: python/setup.py
Comment (tcuongd) [inline python/setup.py:113]: Is this redundant?

File: README.md
Comment (tcuongd) [reply README.md:69]: Hmm since this is specific to Prophet I wonder if we can call it `PROPHET_REPACKAGE_CMDSTAN` and have the default as `True` instead (makes the code easier to read I think).

File: python/setup.py
Comment (tcuongd) [inline python/setup.py:77]: Neat!

File: python/setup.py
Comment (tcuongd) [inline python/setup.py:68]:
```suggestion
        if platform.platform().startswith("Win"):
```
?

**Review by tcuongd — COMMENTED** [review]
This is looking great so far, thanks heaps. I've just done a first pass and haven't looked at the build / test results just yet, but will do that soon. Just left a few questions / suggestions, mostly small.

**Comment by tcuongd** [comment]
> For the Python 3.7 tests, the specific test test_diagnostics.test_cross_validation fails
> 
> The failure is due to this line:
> 
> params = self.stan_to_dict_numpy( self.stan_fit.column_names, self.stan_fit.optimized_params_np )
> 
> which fails because the values in these variables are incorrectly parsed from the cmdstan CSV output.
> 
> The cmdstan CSV output looks like it did not output completely correctly. It is missing a line-break after the parameter names and their values.
> 
> The CSV has both the parameter names and values on the same line, which is not the format the cmdstanpy expects.
> 
> Sample CSV attached below: [prophet_model-20220211103525.csv](https://github.com/facebook/prophet/files/8050040/prophet_model-20220211103525.csv)
> 
> Environment:
> 
> * Python 3.7
> * Test: test_diagnostics.test_cross_validation
> * Test only fails on method=Newton and fails when the cross validation is parallelized using ThreadPoolExecutor or ProcessPoolExecutor
>   
>   * Possibly a race condition with the csv file names only being unique to the second?
> 

Amazing find, thank you @malmashhadani-88!

File: python/requirements.txt
Comment (akosfurton) [reply python/requirements.txt:3]: There are a few references to pystan in `setup.py` and `models.py`

They appear straightforward to remove. This would mean that not only is `cmdstanpy` the default backend, but also that `pystan` will not be supported at all

File: python/prophet/models.py
Comment (WardBrian) [reply python/prophet/models.py:74]: I was waiting for the other test issues to be resolved before trying on my windows machine. Doing that now.

File: python/requirements.txt
Comment (WardBrian) [reply python/requirements.txt:3]: I can remove PyStan all together, but the main reason the requirements changed here is so I could do local testing and it snuck into a commit. 

I think it would be worth a separate PR to remove pystan, as it can also do a lot of cleanup/simplifying of the code

File: python/prophet/models.py
Comment (WardBrian) [reply python/prophet/models.py:74]: I believe I've located the windows issue

**Comment by akosfurton** [comment]
Looks great to me. 

Agree on the 2 next steps, and then should be ready to merge.

I believe this PR + a follow up PR to remove the PyStan backend should unblock https://github.com/facebook/prophet/issues/2041 

Then, prophet 1.1 can be released to PyPi https://github.com/facebook/prophet/issues/2064

**Comment by WardBrian** [comment]
I'm also happy to put together the PR which removes the pystan dependency. Would the project like to maintain the IStanBackend abstraction or just collapse everything down to the CmdStanBackend class?

**Comment by akosfurton** [comment]
What else is left before we can merge this branch?

**Comment by WardBrian** [comment]
Someone needs to approve the workflow runs to make sure tests do pass (I've confirmed locally on Ubuntu and Windows, but still good to check)

I will then do one final commit which re-adds pystan to requirements and turns off the wheel building GHA for PRs, at which point it can be merged.

**Comment by WardBrian** [comment]
Hi @tcuongd - sorry for the silly error. Would you be able to run the workflows again now?

**Comment by WardBrian** [comment]
Looks like the most recent run failed on a duplicate copy of the windows build, but suceeded on the main one!
https://github.com/facebook/prophet/actions/runs/1853320460

I've removed the duplicate and we should be good to go. It's currently only building wheels for py36,37, and 38 - I assume that 39 and 3.10 will also be desired?

**Comment by WardBrian** [comment]
I can confirm the Linux wheel works on Ubuntu 20.04, Python 3.8.12, _without CmdStan installed separately_

**Comment by WardBrian** [comment]
Running on my fork here: https://github.com/WardBrian/prophet/actions/runs/1855227353

**Comment by WardBrian** [comment]
I've just released adding back pystan will prevent building on newer versions. I'll run it [once](https://github.com/WardBrian/prophet/runs/5223171372?check_suite_focus=true) to test and them go back to only building up to 3.8 in the final commit.

**Comment by akosfurton** [comment]
> I've just released adding back pystan will prevent building on newer versions. I'll run it [once](https://github.com/WardBrian/prophet/runs/5223171372?check_suite_focus=true) to test and them go back to only building up to 3.8 in the final commit.

Good point. You can add 3.9 and 3.10 after removing pystan

**Comment by WardBrian** [comment]
https://github.com/WardBrian/prophet/actions/runs/1855227353#artifacts Wheels for 3.6-3.10 on all 3 platforms. Pushing final commit for this PR now, @tcuongd if you could review I believe this is ready to merge

**Comment by WardBrian** [comment]
@tcuongd - is there anything else you'd like to see done in this PR before merging?

**Comment by akosfurton** [comment]
@bletham , could you review this PR as well? It is ready to merge

**Comment by CJSmith-0141** [comment]
+1 on this, this PR solves an issue I'm having with python 3.10 and prophet

**Comment by ShantanuKumar** [comment]
Hi, I tried to use these changes to train model using prophet but I get this error about missing `libtbb.dylib`

**Comment by WardBrian** [comment]
Can you share the error?

**Comment by ShantanuKumar** [comment]
This is generated on macos. I was able to fix it inside a docker container with linux by manually installing `tbb`. I tried to do the same on macos using `brew install tbb` but that doesn't work.
```
INFO:prophet:Disabling yearly seasonality. Run prophet with yearly_seasonality=True to override this.
INFO:prophet:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.
INFO:cmdstanpy:start chain 1
INFO:cmdstanpy:finish chain 1
WARNING:prophet.models:Optimization terminated abnormally. Falling back to Newton.
INFO:cmdstanpy:start chain 1
INFO:cmdstanpy:finish chain 1
Traceback (most recent call last):
  File "/Users/developer/.pyenv/versions/abb-crc-env/lib/python3.9/site-packages/prophet/models.py", line 105, in fit
    self.stan_fit = self.model.optimize(**args)
  File "/Users/developer/.pyenv/versions/abb-crc-env/lib/python3.9/site-packages/cmdstanpy/model.py", line 435, in optimize
    raise RuntimeError(msg)
RuntimeError: Error during optimization.
RunSet: chains=1
 cmd:
        ['/Users/developer/.pyenv/versions/3.9.0/envs/abb-crc-env/lib/python3.9/site-packages/prophet/stan_model/prophet_model.bin', 'random', 'seed=36598', 'data', 'file=/var/folders/9w/9_4n35r57d707zkrk86rvh9w0000gn/T/tmpffcs4eo7/j7fjjyax.json', 'init=/var/folders/9w/9_4n35r57d707zkrk86rvh9w0000gn/T/tmpffcs4eo7/rs05hk_o.json', 'output', 'file=/private/var/folders/9w/9_4n35r57d707zkrk86rvh9w0000gn/T/tmpxll2ozny/prophet_model-202203180948-1.csv', 'method=optimize', 'algorithm=lbfgs', 'iter=10000']
 retcodes=[-6]
 error_msgs:
        /private/var/folders/9w/9_4n35r57d707zkrk86rvh9w0000gn/T/tmpxll2ozny/prophet_model-202203180948-1-stderr.txt
chain_id 1:
dyld[30142]: Library not loaded: @rpath/libtbb.dylib
  Referenced from: /Users/developer/.pyenv/versions/3.9.0/envs/abb-crc-env/lib/python3.9/site-packages/prophet/stan_model/prophet_model.bin
  Reason: tried: '/private/var/folders/9w/9_4n35r57d707zkrk86rvh9w0000gn/T/pip-install-oqo49qox/prophet_deddf0531f6c41cf8980f3b9563b7023/python/build/lib.macosx-11.5-x86_64-3.9/prophet/stan_model/cmdstan-2.26.1/stan/lib/stan_math/lib/tbb/libtbb.dylib' (no such file), '/private/var/folders/9w/9_4n35r57d707zkrk86rvh9w0000gn/T/pip-install-oqo49qox/prophet_deddf0531f6c41cf8980f3b9563b7023/python/build/lib.macosx-11.5-x86_64-3.9/prophet/stan_model/cmdstan-2.26.1/stan/lib/stan_math/lib/tbb/libtbb.dylib' (no such file), '/usr/local/lib/libtbb.dylib' (no such file), '/usr/lib/libtbb.dylib' (no such file)
```

**Comment by WardBrian** [comment]
I will investigate this. In the mean time, you may be able to use `brew install tbb@2020` to get it working

**Comment by WardBrian** [comment]
@ShantanuKumar were you using the downloads from the link I posted above (https://github.com/WardBrian/prophet/actions/runs/1855227353)?

**Comment by ShantanuKumar** [comment]
@WardBrian I build the library locally from the commit without `pystan` on your forked branch.
```
pip install git+https://github.com/WardBrian/prophet.git@18f19f7f989bc38489a1601fa111193f2772dce1#subdirectory=python
```

**Comment by WardBrian** [comment]
Are you able to download the correct wheel from that GithubAction run and try that?

**Comment by ShantanuKumar** [comment]
I only see wheels for 3.8, but I am using python 3.9

**Comment by WardBrian** [comment]
If you download the zip file you will find wheels for many versions

**Comment by ShantanuKumar** [comment]
That worked!

**Comment by ShantanuKumar** [comment]
Can we please merge this?

**Comment by tcuongd** [comment]
Hmm not sure why CI hasn't been running for the more recent pushes, it ran here: https://github.com/facebook/prophet/actions/runs/1853320459
I'd like to kick off another CI check before merging; @WardBrian would you be able to give me push access to your fork? I can usually get CI working again by making a push from my account

**Review by tcuongd — APPROVED** [review]
Amazing stuff!!! 

* Great to see that `cmdstanpy.install_cmdstan()` is fast now and we don't need to use `wget`; makes the wheel build workflow so much cleaner.
* The refactor to `setup.py` looks all good to me, again much cleaner when we make use of the `cmdstanpy` internals.
* Thanks so much to everyone for figuring out tricky bugs like the interaction between multiprocessing and writing to csvs, and the Windows TBB stuff.

Happy to merge this PR but please keep in mind:
* I'm unable to release a new package version to PyPI, so for now Prophet on PyPI will remain at 1.0.1, with all of the pystan installation code, and will only support Python 3.6-3.8.
* The changes in this PR will be accessible by cloning the Prophet repo and building the package manually from source (this is effectively the 'development version', which I've set to 1.1). I've added a section for how to do this in the README. (ps: @WardBrian could you have a read of these changes and let me know if it makes sense) 
* If you don't want to build from source, but want to use Prophet without worrying about `pystan`, please download the latest wheels here: https://github.com/WardBrian/prophet/actions/runs/1855227353 -- the downloaded wheel files can be installed using `pip`.
 
I understand this is confusing but 🤞 we can get a PyPI release soon (then everyone will _actually_ be able to "just pip install prophet", heh).

**Comment by thomasaarholt** [comment]
(been following this PR with quite some interest)
What is the blocker for a PyPi release?

**Comment by tcuongd** [comment]
> (been following this PR with quite some interest) What is the blocker for a PyPi release?

I think Ben (one of the package authors) has the release process set up on his machine, but I haven't been able to get in touch with him recently.

**Comment by WardBrian** [comment]
@tcuongd the README changes look good to me. I'm actually not 100% sure if the final section ("Using `cmdstanpy` with Windows requires a Unix-compatible C compiler such as mingw-gcc.") is true with the pre-compiled installations from wheels, but it certainly is if building from source.

Shame about the PyPi releases. This is part of why on CmdStanPy we do releases through and [action](https://github.com/stan-dev/cmdstanpy/blob/cc6daf27c0d5032eb3dbb0204df5991e32f64210/.github/workflows/release.yml#L93)

**Pull request merged**
